File: src/test/java/net/jsourcerer/webdriver/jserrorcollector/SimpleTest.java
package net.jsourcerer.webdriver.jserrorcollector;

import static org.junit.Assert.assertEquals;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;

import org.junit.Test;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxProfile;


/**
 * 
 * @author Marc Guillemot
 * @version $Revision:  $
 */
public class SimpleTest {
	private final String urlSimpleHtml = getResource("simple.html");
	private final JavaScriptError errorSimpleHtml = new JavaScriptError("TypeError: null has no properties", urlSimpleHtml, 9, null);
	
	private final String urlWithNestedFrameHtml = getResource("withNestedFrame.html");
	private final JavaScriptError errorWithNestedFrameHtml = new JavaScriptError("TypeError: \"foo\".notHere is not a function", urlWithNestedFrameHtml, 7, null);

	private final String urlWithPopupHtml = getResource("withPopup.html");
	private final String urlPopupHtml = getResource("popup.html");
	private final JavaScriptError errorPopupHtml = new JavaScriptError("ReferenceError: error is not defined", urlPopupHtml, 5, null);

	private final String urlWithExternalJs = getResource("withExternalJs.html");
	private final String urlExternalJs = getResource("external.js");
	private final JavaScriptError errorExternalJs = new JavaScriptError("TypeError: document.notExisting is undefined", urlExternalJs, 1, null);

	/**
	 *
	 */
	@Test
	public void simple() throws Exception {
		final WebDriver driver = buildFFDriver();
		try{
			driver.get(urlSimpleHtml);
			
			final List<JavaScriptError> expectedErrors = Arrays.asList(errorSimpleHtml);
			final List<JavaScriptError> jsErrors = JavaScriptError.readErrors(driver);
			assertEquals(expectedErrors, jsErrors);
		}finally{
			driver.quit();
		}
	}
	
	/**
	 *
	 */
	@Test
	public void errorInNestedFrame() throws Exception {
		final List<JavaScriptError> expectedErrors = Arrays.asList(errorWithNestedFrameHtml, errorSimpleHtml);

		final WebDriver driver = buildFFDriver();
		try{
			driver.get(urlWithNestedFrameHtml);
			
			final List<JavaScriptError> jsErrors = JavaScriptError.readErrors(driver);
			assertEquals(expectedErrors.toString(), jsErrors.toString());
		}finally{
			driver.quit();
		}
	}
	
	/**
	 *
	 */
	@Test
	public void errorInPopup() throws Exception {
		final List<JavaScriptError> expectedErrors = Arrays.asList(errorPopupHtml);

		final WebDriver driver = buildFFDriver();
		try{
			driver.get(urlWithPopupHtml);
			driver.findElement(By.tagName("button")).click();
			driver.switchTo().window("The Popup");
	
			final List<JavaScriptError> jsErrors = JavaScriptError.readErrors(driver);
			assertEquals(expectedErrors.toString(), jsErrors.toString());
		}finally{
			driver.quit();
		}
	}

	/**
	 *
	 */
	@Test
	public void errorInExternalJS() throws Exception {
		final List<JavaScriptError> expectedErrors = Arrays.asList(errorExternalJs);

		final WebDriver driver = buildFFDriver();
		try{
			driver.get(urlWithExternalJs);
	
			final List<JavaScriptError> jsErrors = JavaScriptError.readErrors(driver);
			assertEquals(expectedErrors.toString(), jsErrors.toString());
		}finally{
			driver.quit();
		}
	}

	WebDriver buildFFDriver() throws IOException {
		FirefoxProfile ffProfile = new FirefoxProfile();
		ffProfile.addExtension(new File("firefox")); // assuming that the test is started in project's root
		return new FirefoxDriver(ffProfile);
	}

	private String getResource(final String string) {
		String resource = getClass().getClassLoader().getResource(string).toExternalForm();
		if (resource.startsWith("file:/") && !resource.startsWith("file:///")) {
			resource = "file://" + resource.substring(5);
		}
		return resource;
	}
}

File: dotnet/JSErrorCollector.Tests/SimpleTest.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.IO;
using OpenQA.Selenium;
using OpenQA.Selenium.Firefox;
using System.Collections.Generic;

namespace JSErrorCollector.Tests
{
    [TestClass]
    public class SimpleTest
    {
        public TestContext TestContext { get; set; }

        private string urlSimpleHtml;
        private JavaScriptError errorSimpleHtml;

        private string urlWithNestedFrameHtml;
        private JavaScriptError errorWithNestedFrameHtml;

        private string urlWithPopupHtml;
        private string urlPopupHtml;
        private JavaScriptError errorPopupHtml;

        private string urlWithExternalJs;
        private string urlExternalJs;
        private JavaScriptError errorExternalJs;

        [TestInitialize]
        public void Initialize()
        {
            urlSimpleHtml = GetResource("simple.html");
            errorSimpleHtml = new JavaScriptError("TypeError: null has no properties", urlSimpleHtml, 9);

            urlWithNestedFrameHtml = GetResource("withNestedFrame.html");
            errorWithNestedFrameHtml = new JavaScriptError("TypeError: \"foo\".notHere is not a function", urlWithNestedFrameHtml, 7);

            urlWithPopupHtml = GetResource("withPopup.html");
            urlPopupHtml = GetResource("popup.html");
            errorPopupHtml = new JavaScriptError("ReferenceError: error is not defined", urlPopupHtml, 5);

            urlWithExternalJs = GetResource("withExternalJs.html");
            urlExternalJs = GetResource("external.js");
            errorExternalJs = new JavaScriptError("TypeError: document.notExisting is undefined", urlExternalJs, 1);
        }

        /**
	 *
	 */
        [TestMethod]
        public void Simple()
        {
            using (IWebDriver driver = BuildFFDriver())
            {
                driver.Navigate().GoToUrl(urlSimpleHtml);

                IEnumerable<JavaScriptError> expectedErrors = new List<JavaScriptError>() { errorSimpleHtml };
                IEnumerable<JavaScriptError> jsErrors = JavaScriptError.ReadErrors(driver);
                AssertErrorsEqual(expectedErrors, jsErrors);
            }
        }

        /**
         *
         */
        [TestMethod]
        public void ErrorInNestedFrame()
        {
            IEnumerable<JavaScriptError> expectedErrors = new List<JavaScriptError>() { errorWithNestedFrameHtml, errorSimpleHtml };

            using (IWebDriver driver = BuildFFDriver())
            {
                driver.Navigate().GoToUrl(urlWithNestedFrameHtml);

                IEnumerable<JavaScriptError> jsErrors = JavaScriptError.ReadErrors(driver);
                AssertErrorsEqual(expectedErrors, jsErrors);
            }
        }

        /**
         *
         */
        [TestMethod]
        public void ErrorInPopup()
        {
            IEnumerable<JavaScriptError> expectedErrors = new List<JavaScriptError>() { errorPopupHtml };

            using (IWebDriver driver = BuildFFDriver())
            {
                driver.Navigate().GoToUrl(urlWithPopupHtml);
                driver.FindElement(By.TagName("button")).Click();

                IEnumerable<JavaScriptError> jsErrors = JavaScriptError.ReadErrors(driver);
                AssertErrorsEqual(expectedErrors, jsErrors);
            }
        }

        /**
         *
         */
        [TestMethod]
        public void ErrorInExternalJS()
        {
            IEnumerable<JavaScriptError> expectedErrors = new List<JavaScriptError>() { errorExternalJs };

            using (IWebDriver driver = BuildFFDriver())
            {
                driver.Navigate().GoToUrl(urlWithExternalJs);

                IEnumerable<JavaScriptError> jsErrors = JavaScriptError.ReadErrors(driver);
                AssertErrorsEqual(expectedErrors, jsErrors);
            }
        }

        private void AssertErrorsEqual(IEnumerable<JavaScriptError> expectedErrors, IEnumerable<JavaScriptError> actualErrors)
        {
            string expected = "";
            foreach (JavaScriptError err in expectedErrors)
            {
                expected += err.ErrorMessage + " (line " + err.LineNumber + ")";
            }

            string actual = "";
            foreach (JavaScriptError err in actualErrors)
            {
                actual += err.ErrorMessage + " (line " + err.LineNumber + ")";
            }

            Assert.AreEqual(expected, actual);
        }


        private IWebDriver BuildFFDriver()
        {
            FirefoxProfile ffProfile = new FirefoxProfile();
            
            string xpiPath;
            int depth = 0;
            do
            {
                string relative = "";
                for (int i = 0; i < depth; i++)
                    relative += @"..\";

                xpiPath = Path.Combine(TestContext.DeploymentDirectory, relative, @"dist\JSErrorCollector.xpi");

                depth++;
                if (depth > 10)
                    throw new FileNotFoundException("Could not find JSErrorCollector.xpi after 10 iterations.");

            } while (!File.Exists(xpiPath));

            ffProfile.AddExtension(xpiPath);

            return new FirefoxDriver(ffProfile);
        }

        private String GetResource(String fileName)
        {
            string resourceContent = TestResources.ResourceManager.GetString(Path.GetFileNameWithoutExtension(fileName));

            string tempResourcePath = Path.Combine(TestContext.DeploymentDirectory, fileName);
            using (StreamWriter sw = new StreamWriter(tempResourcePath))
            {
                sw.Write(resourceContent);
            }

            string resourceUrl = "file://" + tempResourcePath.Replace(Path.DirectorySeparatorChar, '/');
		    return resourceUrl;
	    }
    }
}
